tests/metrics_tests.cpp

```diff
@@ -0,0 +1,718 @@
+TEST_F(MetricsTest, RecoverSuccessWithLegacyState)
+{
+  MockMetricsService metricsService;
+
+  const hashmap<string, int> containers =
+    {{"container1", 1234}, {"container2", 5678}};
+
+  LegacyStateStore stateStore(containers, statePath);
+
+  // Check that the container state directory exists. We will confirm that it
+  // has been moved after recovery.
+  ASSERT_TRUE(os::exists(path::join(statePath, "containers")));
+
+  // In the recovery case, the metrics module doesn't use the returned host and
+  // port for anything. We inject some dummy values here to avoid addressing the
+  // fact that the order in which the requests is made is not certain.
+  ContainerStartResponse responseBody;
+  responseBody.set_statsd_host("127.0.0.1");
+  responseBody.set_statsd_port(1234);
+
+  process::http::Response response(
+      string(jsonify(JSON::Protobuf(responseBody))),
+      process::http::Status::CREATED,
+      "application/json");
+
+  Future<process::http::Request> request1;
+  Future<process::http::Request> request2;
+  EXPECT_CALL(*metricsService.mock, container(_))
+    .WillOnce(DoAll(FutureArg<0>(&request1),
+                    Return(response)))
+    .WillOnce(DoAll(FutureArg<0>(&request2),
+                    Return(response)));
+
+  // Construct the agent's recovered state, which is passed
+  // into the module's `recover()` method.
+  vector<mesos::slave::ContainerState> agentState;
+
+  foreachkey (const string& containerId_, containers) {
+    ContainerID containerId;
+    containerId.set_value(containerId_);
+
+    mesos::slave::ContainerState containerState;
+
+    containerState.mutable_container_id()->CopyFrom(containerId);
+    containerState.set_pid(0);
+    containerState.set_directory(os::getcwd());
+
+    agentState.push_back(containerState);
+  }
+
+  Future<Nothing> recovered = isolator->recover(agentState, {});
+
+  // Verify the contents of the module's requests.
+  vector<Future<process::http::Request>> requests{request1, request2};
+  foreach (const Future<process::http::Request>& request, requests) {
+    AWAIT_READY(request);
+
+    Try<ContainerStartRequest> requestBody =
+      parse<ContainerStartRequest>(request->body);
+
+    ASSERT_SOME(requestBody);
+    ASSERT_EQ(request->method, "POST");
+
+    ASSERT_TRUE(requestBody->has_statsd_host());
+    ASSERT_EQ(requestBody->statsd_host(), LEGACY_HOST);
+
+    ASSERT_TRUE(requestBody->has_statsd_port());
+    ASSERT_TRUE(containers.contains(requestBody->container_id()));
+    ASSERT_EQ(
+        requestBody->statsd_port(),
+        containers.at(requestBody->container_id()));
+  }
+
+  AWAIT_READY(recovered);
+
+  ASSERT_FALSE(os::exists(statePath));
+}
+
+
+// When the isolator's `recover()` method is called, the legacy state directory
+// has been set in the module configuration, but the state directory does not
+// exist, then the module should return a ready future.
+TEST_F(MetricsTest, RecoverSuccessWithoutLegacyDirectory)
+{
+  Clock::pause();
```

Comment:
Some of these tests pause the clock, but do not resume them afterwards.

It doesn't seem like there is very much code reliant on timeouts in this test suite though (and any tests which do exercise timeouts have the clock paused), so it should be fine 🤔 .